Add GhostLowerDElems relationship manager for side user objects

lindsayad opened the pull request:
Also another bug fix in `BoundaryElemIntegrityCheckThread`.

Closes #21687

yjung-anl committed:
additional bug fix in BoundaryElemIntegrityCheckThread

yjung-anl committed:
add LowerDIntegralSideUserObject for reproducing error in SideUserObject for Lower-d variable access

lindsayad committed:
Add GhostLowerDElems relationship manager

Closes #21687

lindsayad commented:
fascinating ... that distributed mesh failure in heat conduction is definitely caused by this PR ... but why?

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1119839/) on 6bfd0da wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21699/site)

This comment will be updated on new commits.

lindsayad committed:
Dont query for node ptr in construction of NodalVariableValue

The node may be deleted during later tasks of simulation setup, such as during
post late geometric ghosting functor addition remote element deletion

lindsayad committed:
Make thermal contact work post remote element deletion

lindsayad commented:
The reason is that the new RM was triggering remote element deletion which hadn't been triggered before. This required beefing up the robustness of some objects with respect to distributed mesh

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1122141/) on ed6ead4 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>1a9cd6</th>
      <th colspan=4>#21699 ed6ead</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.79%</td>
      <td>83.79%</td>
      <td>+0.00%</td>
      <td>99.03%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>78038</td>
      <td>78076</td>
      <td>+38</td>
      <td>102</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15100</td>
      <td>15106</td>
      <td>+6</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21699/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21699/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21699/coverage/xfem)

</details>



This comment will be updated on new commits.

yjung-anl commented:
@lindsayad 
Thank you for addressing my issue.

recuero commented on framework/src/postprocessors/NodalVariableValue.C:
```diff
@@ -33,15 +33,20 @@ NodalVariableValue::NodalVariableValue(const InputParameters & parameters)
+void
+NodalVariableValue::initialSetup()
+{
+  _node_ptr = _mesh.getMesh().query_node_ptr(getParam<unsigned int>("nodeid"));
```
What's the change in this PR that required querying the nodeid at the `initialSetup` step?

lindsayad replied:
before this PR, there were no instances where this object was used in a truly distributed mesh. This PR triggered mesh distribution, e.g. remote element deletion, for at least one simulation in which this object was present. Late mesh distribution (e.g. late mesh element deletion) occurs after user object construction. So the result of `query_elem_ptr` if run at construction time would be a misnomer if later remote elements (and their connected nodes) are deleted

lindsayad commented on framework/src/relationshipmanagers/GhostLowerDElems.C:
```diff
@@ -0,0 +1,78 @@
+    for (const auto s : elem->side_index_range())
+    {
+      const Elem * const neighbor = elem->neighbor_ptr(s);
+      const bool elem_owns_lowerd = !neighbor || elem->id() < neighbor->id();
```
There is storage on `MooseMesh` for querying for a lower-dimensional element from a higher-dimensional element and side. However, we only store a map from higher-d element and side to lower-d element for *one* of the higher dimensional elements on an internal side. So we choose to store with a higher-dimensional element key based on which higher-dimensional element that "shares" the lower-dimensional element has a lower ID. So basically if either

1. there is no neighbor -> this higher dimensional element definitely is the key in the moose mesh map
2. there is a neighbor, but it has a higher ID than us -> we still are the key in the moose mesh map. If the neighbor has the lower ID, then *it* is the key in the moose mesh map

recuero replied:
The logic seems to allow for coupling more than one higher-d neighbor, but not a real case?

lindsayad replied:
I don't think I fully understand the question

recuero replied:
> we only store a map from higher-d element and side to lower-d element for one of the higher dimensional elements on an internal side

How many higher-d elements that have a side defining a lower-d element can we have?

lindsayad replied:
It will be proportional to the number of elements in the mesh. These lower-d elements that the Griffin team uses for the hybrid finite-element method exist on every single internal side in the mesh

lindsayad replied:
For a replicated mesh, the number of keys in the `_higher_d_elem_side_to_lower_d_elem` map will be equal to the number of faces in the mesh

lindsayad commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -1190,20 +1184,28 @@ AutomaticMortarGeneration::buildCouplingInformation()
+    // Lower-LowerDimensionalPrimary
+    coupling_info[secondary_elem->processor_id()].emplace_back(secondary_elem->id(),
+                                                               primary_elem->id());
+    if (secondary_elem->processor_id() != _mesh.processor_id())
+      // We want to keep information for nonlocal lower-dimensional secondary element point
+      // neighbors for mortar nodal aux kernels
+      _mortar_interface_coupling[secondary_elem->id()].insert(primary_elem->id());
```
it wasn't just for mortar nodal aux kernels. We were missing the lower dimensional primary element ghosting in 3D for mortar constraints. We just never noticed it before because we never deleted remote elements. This use case was covered in the heat conduction module, so we do have coverage within the MOOSE repo. However, if you're asking whether we have a framework level test for that yet ... I think the answer is no

recuero replied:
> We were missing the lower dimensional primary element ghosting in 3D for mortar constraints

So the lower-d primary mesh suddenly stops (no ghosting) for parallel 3D applications? I guess we are not hitting this case from the contact module either.

lindsayad replied:
No. For some reason, which I should figure out, even if a user asks for `--distributed-mesh` with mortar they are not getting it in most cases

lindsayad replied:
This is the code in `AugmentSparsityOnInterface::operator()` which is responsible for most distributed mesh simulations not being truly distributed with respect to the elements contributing to mortar
```c++
  if ((!amg || _use_displaced_mesh) && !_is_coupling_functor)
  {
    for (const Elem * const elem : _mesh->active_element_ptr_range())
    {
      if (generating_mesh)
      {
        // We are still generating the mesh, so it's possible we don't even have the right boundary
        // ids created yet! So we actually ghost all boundary elements and all lower dimensional
        // elements who have parents on a boundary
```
`generating_mesh` is `true` while mesh generators are running, which is when `prepare_for_use/delete_remote_elements` gets called on the libMesh mesh(es). The addition of this new `GhostLowerDElems` RM means that in `SetupMeshCompleteAction` we now enter the conditional at the bottom here whenever there are `SideUserObjects` in the simulation (because this new RM is a late geometric ghosting RM):
```c++
  else if (_current_task == "delete_remote_elements_after_late_geometric_ghosting")
  {
    TIME_SECTION("deleteRemoteElems", 2, "Deleting Remote Elements");

    if (_displaced_mesh &&
        (_mesh->needsRemoteElemDeletion() != _displaced_mesh->needsRemoteElemDeletion()))
      mooseError("Our reference and displaced meshes are not in sync with respect to whether we "
                 "should delete remote elements.");

    // We currently only trigger the needsRemoteDeletion flag if somebody has requested a late
    // geometric ghosting functor and/or we have a displaced mesh
    if (_mesh->needsRemoteElemDeletion())
```

recuero replied:
I see. I'd figure the solution isn't as simple as delaying the call to  `AugmentSparsityOnInterface::operator()` even for cases that do not require `SetupMeshCompleteAction`.

lindsayad replied:
the solution to what? Ensuring that we actually delete remote elements for all distributed mesh simulations? Or something else?

recuero replied:
The solution `the mortar mesh becoming truly distributed` when not using `SideUserObjects`.

lindsayad replied:
I will make the geometric ghosting functor in `MortarConsumerInterface::validParams` a late one ... this should ensure that we always get a distributed mortar interface for undisplaced meshes. However for displaced meshes we will still ghost the entire mortar interface since we do not know how the mesh will move and consequently what projections we will have. The alternative to ghosting the entire mortar interface for displaced meshes would be to do communication of elements at each residual/Jacobian evaluation

lindsayad replied:
I'm not sure if I want to take that on right now even if it's a good idea (and I'm not sure whether the memory gains of having more distribution would be worth the communcation cpu costs)

lindsayad replied:
oh dear ... now it looks like there are issues with the VCP :frowning:

lindsayad committed:
Ouput nodal geometry using the Output system

lindsayad committed:
Erase deleted secondary elements from the secondary-to-msm map

This avoids calling methods on deleted elements later on in the
simulation

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 4747096de5d6c69ed79f28e38ce45c76546364c3
+Subproject commit ab868b661940509021b08165c2fa32e8099541f6
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1126868/) on 0e9cb7a wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml
  conda/libmesh/meta.yaml

The libmesh submodule or configuration was changed but the conda build config was not

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1126991/) on 29c485a wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml
  conda/libmesh/meta.yaml

The libmesh submodule or configuration was changed but the conda build config was not

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1127036/) on da17f43 wanted to post the following:

A change of the following file(s) triggered this check:

  conda/libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1127036/) on da17f43 wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The libmesh submodule or configuration was changed but the conda build config was not

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1127036/) on da17f43 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

recuero commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -1397,11 +1397,22 @@ class FEProblemBase : public SubProblem, public Restartable
       const bool correct_edge_dropping,
       const Real minimum_projection_angle);
 
+  ///@{
+  /**
+   * Return the undisplaced or displaced mortar generation object associated with the provided
+   * boundaries and subdomains
+   */
```
I think the comment should be above`///@{`

recuero commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -47,6 +49,159 @@
+  ///@}
+};
+
+registerMooseObject("MooseApp", MortarNodalGeometryOutput);
```
I can't see a test file requesting this output. Is it supposed to?

lindsayad replied:
Any input with `debug_mesh = true` in a mortar constraint will implicitly create this output object. You have one input in the contact module that does this ... otherwise I wouldn't have had to create this hahaha. But yea if we don't see coverage for this in the framework testing, I will add a framework level test requesting debug output

recuero replied:
My understanding: You are creating an output object that's registered. But I don't think a user can leverage it from the input file, despite being registered. It's still driven from the `debug_mesh` flag. May be potentially confusing to the user?

lindsayad replied:
Yea I can see that. I will think about how to do this best without doing any registration

lindsayad committed:
Just use ids in secondary elem to msm map

Then we can safely check whether the element still exists using
`query_elem_ptr`. This removes the need to do anything fancy in
the relationship manager classes around `mesh_reinit` and `delete_remote_elements`.
The issue I ran into with that was that one RM already erased the
iterators from the map and then another RM came along and tried
to do erasures using iterator keys that had been erased/invalidated

lindsayad committed:
Remove reinit'ing of ghosting functors

This allows me to remove the libmesh update commits from here

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1128942/) on 1bfb484 wanted to post the following:

A change of the following file(s) triggered this check:

  conda/libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1128942/) on 1bfb484 wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The libmesh submodule or configuration was changed but the conda build config was not

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1128942/) on 1bfb484 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

lindsayad committed:
Move doxygen comments above grouping

lindsayad committed:
For mortar, delay remote element deletion *and* ensure it happens

lindsayad committed:
Remove MortarNodalGeometryOutput from object registry

lindsayad commented:
@dewenyushu it doesn't appear that the VCP works with a distributed mortar mesh. Is this surprising?

lindsayad committed:
Revert ComputeMortarFunctor change to MooseObject

dewenyushu commented:
> @dewenyushu it doesn't appear that the VCP works with a distributed mortar mesh. Is this surprising?

Humm ... VCP is not tested with distributed mesh so I am very surprised that there are some issues with it.  But I am not sure what exactly may cause the issue - do you know how distributed mesh would change the Jacobian, partitioning of DoFs, etc., in general?

lindsayad commented:
distributed mesh change change element numbering and consequently dof numbering, but your vcp tests in the framework have always run with distributed mesh and passed ... the difference is that now we aren't just running with a `DistributedMesh` ... we are actually running with a distributed mortar mesh, e.g. the mortar interface is not replicated on all processes

lindsayad commented:
Aha I found this comment:
```c++
  // TODO: this might not work for distributed mesh and needs to be improved
  NodeRange * active_nodes = _mesh.getActiveNodeRange();
```
this indeed won't work for distributed meshes

recuero commented:
Bummer. When distributed mesh is enabled in the VCP, it seems the only change required is in the commit https://github.com/idaholab/moose/pull/21699/commits/8348eec83ba5833523a9c78ae507ff3c335d32da.

lindsayad committed:
Error if VCP is run with DistributedMesh

lindsayad commented:
That commit made the VCP tests stop working with "distributed mesh" ... but the quotation is important here. It's good @recuero that you recommended ensuring that we always *actually* distribute the mesh. Because if you hadn't, then it's possible that at some point down the road a user would have used mortar and the VCP in conjunction with some other object that caused the mesh to be truly distributed (e.g. a `SideUserObject` after this PR), and then they would have gotten this relatively obscure error out of PETSc. It's much better that we catch this now and give the user a straightforward error telling them truthfully that the VCP is not currently setup to work with a distributed mesh

lindsayad commented:
Ok I think all current review comments should be addressed. Let's see how testing goes!

lindsayad committed:
Add GhostHigherDLowerDPointNeighbors relationship manager

recuero commented on framework/src/relationshipmanagers/GhostHigherDLowerDPointNeighbors.C:
```diff
@@ -0,0 +1,96 @@
+{
+  mooseAssert(_moose_mesh,
+              "The MOOSE mesh must be non-null in order for this relationship manager to work.");
+  const auto & node_to_elem_map = _moose_mesh->nodeToElemMap();
```
How do we know in this map whether it's a lowerD-element or higherD-element node neighbor? (It's probably the wrong question, but it's showing what I don't understand about the coupled ghosting objects).

lindsayad replied:
we don't know *a priori*. The map will map from node to both higher and lower-d elements that use that node. That's why later in the code I check the dimension of the elements that we get from the node key

recuero replied:
Got it.

recuero commented on framework/include/constraints/AutomaticMortarGeneration.h:
```diff
@@ -315,7 +310,7 @@ class AutomaticMortarGeneration : public ConsoleStreamInterface
   /**
    * @return the lower dimensional secondary elements and their associated mortar segment elements
```
Obsolete comment?

recuero commented on framework/src/outputs/OutputWarehouse.C:
```diff
@@ -90,7 +90,7 @@ OutputWarehouse::subdomainSetup()
 
 void
-OutputWarehouse::addOutput(std::shared_ptr<Output> & output)
+OutputWarehouse::addOutput(std::shared_ptr<Output> const output)
```
Why the `const` point qualifier not in the header?

lindsayad replied:
If at some point a developer wanted to intentionally modify this argument, the could remove the `const` in the implementation (.C) file without having to modify the header, causing recompilation of all the translation units that depend on it. As it does not matter to the caller whether this is marked as `const` or not (since it is pass-by-value) it really does not matter whether, from an API standpoint, you mark it as such. However, marking it `const` in the implementation helps to make sure you don't accidentally do something stupid in the implementation, like change `output` to point to some other object

recuero replied:
> As it does not matter to the caller whether this is marked as const or not (since it is pass-by-value) it really does not matter whether

👍 
Thanks for the details.

recuero commented on framework/src/actions/SetupMeshCompleteAction.C:
```diff
@@ -100,6 +101,9 @@ SetupMeshCompleteAction::act()
     // geometric ghosting functor and/or we have a displaced mesh
     if (_mesh->needsRemoteElemDeletion())
```
How is this flag different from a hypothetical `distributedMesh()` flag?

lindsayad replied:
This does raise a good point ... there are two different APIs on the libMesh `MeshBase` object that tell you different things. There is `is_replicated` and  `is_serial`. The former is an indication of the class type. The latter is in indication of the current state of the class. For a `DistributedMesh`, `is_replicated` will always return `false`. However, `is_serial` will return `true` unless a user has explicitly called `set_distributed()` or has had `delete_remote_elements` called. So ... I should check the status of `is_serial` here. If `needsRemoteElemDeletion` is `true`, then I would hope that `is_serial` is also `true`. 

If `is_serial` is `true` and `needsRemoteElemDeletion` is `false` ... then if I try to `delete_remote_elements` I could possibly run across a whole slew of additional cases where previously we have not been running with a distributed mesh when we thought we were ... ok I'm going to try deleting remote elements whenever `!is_replicated() && is_serial()` and see what happens

lindsayad replied:
Sigh, this leads to 4 new failures with distributed mesh

recuero replied:
Would that be because we are not hitting the `is_serial()` flag? I.e. the state isn't the one we want at that point?

lindsayad replied:
No it's because there is likely more MOOSE code that doesn't actually properly support distributed mesh yet

lindsayad replied:
Cross-referencing: created #21798 based on this conversation

recuero commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -912,16 +912,6 @@ FEProblemBase::initialSetup()
-  // during residual and Jacobian evaluation because when displacements are solution variables
-  // the mortar mesh will move and change during the course of a non-linear solve. We DO NOT
-  // redo ghosting during non-linear solve, so for purpose 1) the below call has to be made
-  updateMortarMesh();
```
What's the implication of moving `updateMortarMesh()` from `FEProblemBase::initialSetup()` to `FEProblemBase::init()` ?

recuero reviewed (commented):
Just some questions.

lindsayad committed:
Update doxygen comment for seconaries to msm elems map

recuero reviewed (approved):
Looks good to me.

yjung-anl commented:
@lindsayad  Thank you!!

lindsayad commented:
No problem! Thanks for the commits and bug report!

On Mon, Aug 8, 2022 at 8:38 AM Yeon Sang Jung ***@***.***>
wrote:

> @lindsayad <https://github.com/lindsayad> Thank you!!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/21699#issuecomment-1208288253>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOGA4GSYJM65V26FDCWPSLVYESWFANCNFSM54XL2IAQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
